File: packages/python/src/llm_jury/classifiers/function_adapter.py
from __future__ import annotations

import inspect
from typing import Awaitable, Callable

from .base import ClassificationResult, Classifier


class FunctionClassifier(Classifier):
    def __init__(
        self,
        fn: Callable[[str], tuple[str, float]] | Callable[[str], Awaitable[tuple[str, float]]],
        labels: list[str],
    ) -> None:
        self.fn = fn
        self.labels = labels

    async def classify(self, text: str) -> ClassificationResult:
        result = self.fn(text)
        if inspect.isawaitable(result):
            label, confidence = await result
        else:
            label, confidence = result

        return ClassificationResult(
            label=label,
            confidence=float(confidence),
            raw_output={"label": label, "confidence": float(confidence)},
        )

File: packages/typescript/src/classifiers/functionAdapter.ts
import { defaultClassifyBatch } from "./base.ts";
import type { ClassificationResult, Classifier } from "./base.ts";

export class FunctionClassifier implements Classifier {
  public labels: string[];
  private fn: (text: string) => [string, number] | Promise<[string, number]>;

  constructor(fn: (text: string) => [string, number] | Promise<[string, number]>, labels: string[]) {
    this.fn = fn;
    this.labels = labels;
  }

  async classify(text: string): Promise<ClassificationResult> {
    const [label, confidence] = await this.fn(text);
    // A local function makes no paid call, so its cost is a known zero.
    return {
      label,
      confidence,
      rawOutput: { label, confidence },
      costUsd: 0,
    };
  }

  async classifyBatch(texts: string[]): Promise<ClassificationResult[]> {
    return defaultClassifyBatch(this, texts);
  }
}
